File: Year 1 - Semester 1/Fundamentals of Programming/Labs/FP/src/seminar/group_911/seminar_11/repository/Repository.py
from seminar.group_911.seminar_11.repository.RepositoryException import RepositoryException


class Repository:
    """
    Repository for storing IDObject instances
    """

    def __init__(self):
        self._objects = []


    def store(self, obj):
        if self.find(obj.id) is not None:
            raise RepositoryException("Element having id=" + str(obj.id) + " already stored!")
        self._objects.append(obj)


    def update(self, object1, object2):
        """
        Update the instance given as parameter. The provided instance replaces the one having the same ID
        object - The object that will be updated
        Raises RepositoryException in case the object is not contained within the repository
        """
        el = self.find(object1.id)
        if el is None:
            raise RepositoryException("Element not found!")
        idx = self._objects.index(el)
        self._objects.remove(el)
        self._objects.insert(idx, object2)


    def find(self, objectId):
        for e in self._objects:
            if objectId == e.id:
                return e
        return None


    def delete(self, objectId):
        """
        Remove the object with given objectId from repository
        objectId - The objectId that will be removed
        Returns the object that was removed
        Raises RepositoryException if object with given objectId is not contained in the repository
        """
        object = self.find(objectId)
        if object is None:
            raise RepositoryException("Element not in repository!")
        self._objects.remove(object)
        return object


    def getAll(self):
        return self._objects;


    def __len__(self):
        return len(self._objects)


    def __str__(self):
        r = ""
        for e in self._objects:
            r += str(e)
            r += "\n"
        return r

File: Year 1 - Semester 2/OOP/Lab/T1/repository.cpp
#include "repository.h"

Repository::Repository(){}

Repository::Repository(std::vector<Protein>& _v) : v{_v} {}

void Repository::add(const Protein& p)
{
	if (search(p) != -1){
		throw std::exception("Protein already exists!");
	}
	v.push_back(p);
}

void Repository::remove(const Protein& p)
{
	int pos = search(p);
	if (pos == -1){
		throw std::exception("Protein not found!");
	}
	auto it = std::remove_if(v.begin(), v.end(), [&p](const Protein& p1) {
		return p1 == p; 
		});
	v.erase(it, v.end());
}

int Repository::search(const Protein& p)
{
	auto it = std::find(v.begin(), v.end(), p);
	if (it == v.end())
		return -1;
	return it - v.begin();
}

std::vector<Protein>& Repository::getAll()
{
	return this->v;
}
